packages/react-sdk/src/App.tsx

```diff
@@ -15,20 +15,105 @@
  */
 
 import { useWidgetApi } from '@matrix-widget-toolkit/react';
+import { getLogger } from 'loglevel';
+import { useCallback, useEffect } from 'react';
 import { useTranslation } from 'react-i18next';
 import { Layout, LayoutProps } from './components/Layout';
 import { PageLoader } from './components/common/PageLoader';
-import { useOwnedWhiteboard, useWhiteboardManager } from './state';
+import {
+  isValidWhiteboardDocumentSnapshot,
+  useActiveWhiteboardInstance,
+  useOwnedWhiteboard,
+  useWhiteboardManager,
+} from './state';
+import {
+  cancelableSnapshotTimer,
+  usePersistOnJoinOrInvite,
+} from './state/usePersistOnJoinOrInvite';
+import { useGetDocumentSnapshotQuery } from './store/api';
 
 export type AppProps = {
   layoutProps?: LayoutProps;
 };
 
+const logger = getLogger('App');
+
 export const App = ({ layoutProps }: AppProps) => {
   const { t } = useTranslation('neoboard');
   const { value, loading } = useOwnedWhiteboard();
   const whiteboardManager = useWhiteboardManager();
-  const ownUserId = useWidgetApi().widgetParameters.userId;
+  const whiteboardInstance = useActiveWhiteboardInstance();
+  const widgetApi = useWidgetApi();
+  const ownUserId = widgetApi.widgetParameters.userId;
+  const { limitedHistoryVisibility, delayedPersist, lastMembershipEventTs } =
+    usePersistOnJoinOrInvite();
+  const { data: latestSnapshot } = useGetDocumentSnapshotQuery({
+    documentId: whiteboardInstance.getDocumentId(),
+    validator: isValidWhiteboardDocumentSnapshot,
+  });
+
+  const handlePersist = useCallback(() => {
```

Comment:
Would be nice if the entire logic is split out, e.g. into a hook or function. Otherwise `App` becomes really cluttered.